### PR TITLE
crossdock: Add ctxpropagation behavior

### PR DESCRIPTION
**Pull request by abhinav**
This behavior tests that we propagate baggage consistently across multiple
hops.

This behavior is currently tested against Go only.

---

Depends on https://github.com/yarpc/yarpc-go/pull/207

### CODE REVIEW

File: crossdock/client/ctxpropagation/behavior.go [hunk 1]
```diff
@@ -0,0 +1,303 @@
...
+
+type handler interface {
+	Inject(json.Client, server.TransportConfig) // this is hacky but whatever
+	Handle(*json.ReqMeta, interface{}) (interface{}, *json.ResMeta, error)
+}
+
+func assertBaggageMatches(t crossdock.T, ctx context.Context, want transport.Headers) bool {
+	assert := crossdock.Assert(t)
+	got := baggage.FromContext(ctx)
+	if len(want) == 0 { // nils
```
Comment (breerly) [inline crossdock/client/ctxpropagation/behavior.go]: what do you mean `// nils`?

Comment (abhinav) [inline crossdock/client/ctxpropagation/behavior.go]: That we're doing a `len` check here to gracefully handle `nil` maps. `reflect.DeepEqual(nil, transport.Headers{}) == false`, but `len(nil) == len(transport.Headers{})`.

I'll update the comment.

File: crossdock/client/ctxpropagation/behavior.go [hunk 2]
```diff
@@ -0,0 +1,303 @@
...
+//
+// Behavior parameters:
+// - server: Address of the crossdock test subject server.
+// - transport: The transport to make requests to the test subject with.
+//
+// This behavior sets up yet another server in-process which the Phone
+// procedure on the test subject is responsible for calling. Usually this
+// behavior will receive requests from the Phone procedure on a transport
+// different from the one used to call Phone.
+func Run(t crossdock.T) {
```
Comment (breerly) [inline crossdock/client/ctxpropagation/behavior.go]: nit - this should probably go at the top of the file

File: crossdock/client/ctxpropagation/behavior.go [hunk 3]
```diff
@@ -0,0 +1,303 @@
...
+	"github.com/yarpc/yarpc-go/transport"
+	ht "github.com/yarpc/yarpc-go/transport/http"
+	tch "github.com/yarpc/yarpc-go/transport/tchannel"
+
+	"github.com/uber/tchannel-go"
+	"golang.org/x/net/context"
+)
+
+type handler interface {
+	Inject(json.Client, server.TransportConfig) // this is hacky but whatever
```
Comment (breerly) [inline crossdock/client/ctxpropagation/behavior.go]: This function is really funky - suggest `SetClient` and `SetTransportConfig` instead.

File: crossdock/client/ctxpropagation/behavior.go [hunk 4]
```diff
@@ -0,0 +1,303 @@
...
+				"hello": &singleHopHandler{
+					t:           t,
+					wantBaggage: transport.Headers{"token": "42"},
+				},
+			},
+		},
+		{
+			desc:      "add baggage",
+			procedure: "one",
+			handlers: map[string]handler{
```
Comment (breerly) [inline crossdock/client/ctxpropagation/behavior.go]: I'm in <3 with this test

File: crossdock/main_test.go [hunk 5]
```diff
@@ -95,6 +95,16 @@ func TestCrossdock(t *testing.T) {
 				"transport": []string{"http", "tchannel"},
 			},
 		},
+		{
+			name: "ctxpropagation",
+			axes: axes{
+				"transport": []string{"http", "tchannel"},
```
Comment (breerly) [inline crossdock/main_test.go]: I couldn't tell from reading the test - do we every criss-cross transports in the same test? I want to make sure we aren't breaking the chain...

Comment (abhinav) [inline crossdock/main_test.go]: Yeah, we criss-cross. `transport` specifies the transport we use to talk to the subject, and the other transport is used for the subject to talk to the client.

**Comment by breerly** [comment]
🐑  it.

(fwiw, i do think you should heed the comment about the `Inject` function)

File: crossdock/client/ctxpropagation/behavior.go (hunk 2 above)
Comment (abhinav) [inline crossdock/client/ctxpropagation/behavior.go]: 👍

File: crossdock/client/ctxpropagation/behavior.go (hunk 3 above)
Comment (abhinav) [inline crossdock/client/ctxpropagation/behavior.go]: 👍

File: crossdock/main_test.go (hunk 5 above)
Comment (abhinav) [inline crossdock/main_test.go]: Made the documentation of the behavior a bit clearer.

**Pull request merged**